kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/WatchListDeletable.java

```diff
@@ -16,13 +16,30 @@
 package io.fabric8.kubernetes.client.dsl;
 
 import io.fabric8.kubernetes.client.GracePeriodConfigurable;
+import io.fabric8.kubernetes.client.KubernetesClientTimeoutException;
 import io.fabric8.kubernetes.client.PropagationPolicyConfigurable;
 import io.fabric8.kubernetes.client.Watcher;
 
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+
 public interface WatchListDeletable<T, L> extends Watchable<Watcher<T>>, Versionable<WatchAndWaitable<T>>, Listable<L>, Deletable,
                                                         GracePeriodConfigurable<Deletable>,
                                                         PropagationPolicyConfigurable<EditReplacePatchDeletable<T>>,
                                                         StatusUpdatable<T>,
                                                         Informable<T>
 {
+  
+  /**
+   * Wait for the entire list available at this context to satisfy the given predicate.
+   * @param condition the {@link Predicate} to test
+   * @param pollInterval the polling interval to use, a value less than 1 indicates that every event should be tested
+   * @param timeout after which a {@link KubernetesClientTimeoutException} is thrown
+   * @param timeUnit unit of the pollInterval and timeout
+   * @return the list of items after the condition is met
+   * @throws KubernetesClientTimeoutException if time runs out before the condition is met
+   */
+  List<T> waitUntilListCondition(Predicate<List<T>> condition, long pollInterval, long timeout, TimeUnit timeUnit);
```

Comment:
should we provide a
```java
default List<T> waitUntilListCondition(Predicate<List<T>> condition) {
  return waitUntilListCondition(condition, 0, TimeUnit.SECONDS);
}
```
for a non-polling default method?

Reply:
Okay, I'll add:
```
default List<T> waitUntilListCondition(Predicate<List<T>> condition, long timeout, TimeUnit timeUnit) {
  return waitUntilListCondition(condition, 0, timeout, TimeUnit.SECONDS);
}
```

And move the methods onto Informable.

Reply:
I'm not a big fan of multiplying method overrides when they don't provide an obvious value. This leads to higher cognitive load and potential confusion in my opinion.

Reply:
Well, my suggestion is based on usability. For the average user I don't see the added value of being able to specify a polling interval.

Right now, I wouldn't be able to say what's the recommended approach for any given scenario. I suspect that we'll end up getting questions about this.

I think that the reason to add this variable is based on the underlying implementation and the usage or not of informers. So in the end, I guess the question boils down to the question that given informers+cache should provide the exact same behavior, why should we want to use a polling mechanism instead?

Reply:
> So in the end, I guess the subject boils down to the question that given informers+cache should provide the exact same behavior, why should we want to use a polling mechanism instead?

The polling approach was to match the existing logic of the RollingUpdater - it would of course work without polling as well.  The other thoughts as to why a polling option made sense are for cases where a resource is rapidly changing (and therefore testing on every event may not be desired) or where you want the test to have a temporal component - for example, test every 500 ms, but every 5 seconds emit a log or some other more involved action.  We have a scenario where while waiting for a resource to be ready we periodically check for unscheduled pods.

Another view point is that the polling is like a resync, but at a list level.  If that and other handling were added to the ResourceEventHandler - onRefresh(List<T>), onResync(List<T>) - then we don't necessarily need a dsl method as this scenario could be implemented directly on the Informer.

```
Predicate<List<Pod>> predicate = ...
CompletableFuture<Void> future = ...
SharedInformer<Pod> informer = ...pods().inform(null, 1000);
informer.addEventHandler(new ResourceEventHandler<HasMetadata>() {
    public void onRefresh(List<HasMetadata> objs) {
      test(objs);
    }

    public void onResync(List<HasMetadata> objs) {
      test(objs);
    }
   
    public void onAdd(HasMetadata obj) {
      // do something if I want to test every add
      // requires access to the Informer Cache
    }
  
    ...

    void test(List<HasMetadata> objs) {
      // test the predicate, complete the future
    }
});

``` 
A built-in implementation of this ResourceEventHandler could be supplied - that would be somewhat similar to the ReadinessWatcher (although that looks like it's no longer needed).

Reply:
So if it does make sense to have both options, IMHO it is clearer to have the 2 methods, one that implies the non-polling behavior and another one where it's actually expected.
For me it's confusing that a "magic" 0 will disable polling and assume you want the async behavior instead.
Anyway, we can discuss this point in today's meeting.